libstuff/SDeburr.cpp

```diff
@@ -0,0 +1,134 @@
+         *
+         *   1. Remove the prefix bits (110) from the first byte. That gives us 00011.
+         *   2. Remove the prefix bits (10) from the second byte. That gives us 110001.
+         *   3. Concat these together, and the codepoint for ñ is 00011110001 (which, sure enough, is the binary representation of U+00F1).
+         */
+        size_t start = i;
+
+        // Extract data bits from leading byte
+        // The number of prefix bits is 7 - numLeadingOnes, because the leading ones signaling the length of the sequence are always terminated by a 0.
+        // For example, two leading ones for a 2-byte sequence results in a prefix of 110 in the first byte of the sequence.
+        uint32_t codepoint = inputBytes[i] & ((1 << (7 - numLeadingOnes)) - 1);
+        i++;
+
+        // Process continuation bytes. 
+        for (int j = 1; j < numLeadingOnes; j++) {
+            if (countl_one(inputBytes[i]) != 1) {
+                // Invalid continuation byte, stop processing this sequence
+                break;
+            }
+
+            // Shift the existing codepoint over 6 positions to make room for the new bits
+            codepoint <<= 6;
+
+            // Extract lower 6 bits from continuation and concat them to the existing codepoint
+            codepoint = codepoint | (inputBytes[i] & 0b00111111);
+            i++;
+        }
+
+        // Accent marks by themselves (like ´ ` ^) → skip them
+        if (codepoint >= 0x0300 && codepoint <= 0x036F) {
+            continue;
+        }
+
+        // Map the unicode codepoint to ascii
+        const char* mapped = unicodeToAscii(codepoint);
+        if (mapped == nullptr) {
+            // No conversion needed, keep the original character
+            result.append(reinterpret_cast<const char*>(inputBytes + start), i - start);
+        } else if (*mapped) {
+            // Replace with ASCII equivalent (é→e, ß→ss, etc.)
+            result.append(mapped);
+        } else {
+            // Delete this character (accent marks)
+        }
+    }
+    return result;
+}
+
+void SDeburr::registerSQLite(sqlite3* db) {
+    // SQLite UDF: DEBURR(text) → deburred ASCII string.
+    // Behavior: NULL input → NULL, Non-NULL input → deburred ASCII text
+    // Declared deterministic to enable SQLite optimizations
+    auto sqliteDeburr = [](sqlite3_context* ctx, int argc, sqlite3_value** argv) {
+        if (argc != 1) {
+            sqlite3_result_null(ctx);
+            return;
+        }
+        if (sqlite3_value_type(argv[0]) == SQLITE_NULL) {
+            sqlite3_result_null(ctx);
+            return;
+        }
+        const unsigned char* text = sqlite3_value_text(argv[0]);
+        if (!text) {
+            sqlite3_result_null(ctx);
+            return;
+        }
+        string out = SDeburr::deburr(reinterpret_cast<const char*>(text));
```

Comment:
If `SDeburr::deburr` just takes a `const unsigned char*` instead of a string, that simplifies a few things (including avoiding constructing a string with a copy of `text`)

Reply:
consolidating to the other thread